### PR TITLE
TypeScript: Convert `APIProvider`

### LINKED ISSUE #12660: TypeScript: Convert api provider

### PR DESCRIPTION
**Pull request by miina**
## Context
Converts `APIProvider` to TS

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12660

### CODE REVIEW

**Review by spacedmonkey — APPROVED** [review]

File: packages/story-editor/src/app/api/apiProvider.tsx [hunk 1]
```diff
@@ -17,19 +17,22 @@
...
  * Internal dependencies
  */
 import { useConfig } from '../config';
 import Context from './context';
 
-function APIProvider({ children }) {
+interface APIProviderProps {
+  children: ReactNode;
+}
+function APIProvider({ children }: APIProviderProps) {
```
Comment (barklund) [inline packages/story-editor/src/app/api/apiProvider.tsx]: Let's use `PropsWithChildren<{}>` instead. NB: In React 18 it has a default empty type argument, but not in 17 unfortunately.

File: packages/story-editor/src/app/api/apiProvider.tsx [hunk 2]
```diff
@@ -17,19 +17,22 @@
...
   const { apiCallbacks: actions, cdnURL } = useConfig();
-  const pageTemplates = useRef({
+  const pageTemplates = useRef<{ base: Template[] }>({
```
Comment (barklund) [inline packages/story-editor/src/app/api/apiProvider.tsx]: Why is this ref not just a `Template[]`?

Also, if it was, you'd need perhaps need to type it as `useRef<Template[] | null>(...)`.

**Review by barklund — APPROVED** [review]
Looks good 🎯

File: packages/story-editor/src/app/api/apiProvider.tsx (hunk 1 above)
Comment (miina) [reply packages/story-editor/src/app/api/apiProvider.tsx]: We'll have to use `PropsWithChildren<Record<string, never>>` then since `{}` is not allowed.

**Pull request merged**